pkg/restore/pipeline_items.go

```diff
@@ -0,0 +1,310 @@
+// Copyright 2020 PingCAP, Inc. Licensed under Apache-2.0.
+
+package restore
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/kvproto/pkg/backup"
+	"github.com/pingcap/log"
+	"github.com/pingcap/parser/model"
+	"go.uber.org/zap"
+
+	"github.com/pingcap/br/pkg/conn"
+	"github.com/pingcap/br/pkg/glue"
+	"github.com/pingcap/br/pkg/rtree"
+	"github.com/pingcap/br/pkg/utils"
+)
+
+const (
+	defaultBatcherOutputChannelSize = 1024
+)
+
+// CreatedTable is a table is created on restore process,
+// but not yet filled by data.
+type CreatedTable struct {
+	RewriteRule *RewriteRules
+	Table       *model.TableInfo
+	OldTable    *utils.Table
+}
+
+// TableWithRange is a CreatedTable that has been bind to some of key ranges.
+type TableWithRange struct {
+	CreatedTable
+
+	Range []rtree.Range
+}
+
+// Batcher collects ranges to restore and send batching split/ingest request.
+type Batcher struct {
+	cachedTables   []TableWithRange
+	cachedTablesMu *sync.Mutex
+	rewriteRules   *RewriteRules
+
+	ctx context.Context
+	// joiner is for joining the background batch sender.
+	joiner             chan<- struct{}
+	sendErr            chan<- error
+	outCh              chan<- CreatedTable
+	sender             BatchSender
+	BatchSizeThreshold int
+	size               int32
+}
+
+// Exhaust drains all remaining errors in the channel, into a slice of errors.
+func Exhaust(ec <-chan error) []error {
+	out := make([]error, 0, len(ec))
+	select {
+	case err := <-ec:
+		out = append(out, err)
+	default:
+	}
+	return out
+}
+
+// Len calculate the current size of this batcher.
+func (b *Batcher) Len() int {
+	return int(atomic.LoadInt32(&b.size))
+}
+
+// BatchSender is the abstract of how the batcher send a batch.
+type BatchSender interface {
+	// RestoreBatch will backup all ranges and tables
+	RestoreBatch(ranges []rtree.Range, rewriteRules *RewriteRules, tbs []CreatedTable) error
+	Close()
+}
+
+type tikvSender struct {
+	client         *Client
+	updateCh       glue.Progress
+	ctx            context.Context
```

Comment:
[Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it.](https://pkg.go.dev/context?tab=doc)

Reply:
I got it.